### PR TITLE
dbcsr: fix libxsmm discovery

### PR DESCRIPTION
**Pull request by sheepforce**
## Description of changes

Updates LibXSMM for small matrix multiplications to the latest (3 years old) release. The repo was also moved to the libxsmm organisation.

Fixes LibXSMM discovery in the DBCSR library.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by chessai — APPROVED** [review]

File: pkgs/development/libraries/libxsmm/default.nix [hunk 1]
```diff
@@ -35,10 +35,23 @@ stdenv.mkDerivation rec {
...
+      ];
+  in ''
     mkdir -p $dev/lib/pkgconfig
     mv $out/lib/*.pc $dev/lib/pkgconfig
 
+    # Set rpath for all binaries and shared libraries explicitly.
+    # Otherwise the rpath contains /build/source/lib paths
+    for F in $out/bin/* $out/lib/*.so*; do
+      patchelf --set-rpath ${rpath} $F
+    done
```
Comment (markuskowa) [inline pkgs/development/libraries/libxsmm/default.nix:53]:
```suggestion
```
Completely re-writing the rpath is a pretty rough method.

Patching the makefiles is maybe a bit less intrusive?
This should do the job:
```
diff --git a/Makefile b/Makefile
index 15ee6f056..342502a07 100644
--- a/Makefile
+++ b/Makefile
@@ -875,7 +875,7 @@ endif
 .PHONY: generator
 generator: $(BINDIR)/libxsmm_gemm_generator
 $(BINDIR)/libxsmm_gemm_generator: $(BINDIR)/.make $(OBJFILES_GEN_GEMM_BIN) $(OUTDIR)/libxsmmgen.$(LIBEXT)
-       $(LD) -o $@ $(OBJFILES_GEN_GEMM_BIN) $(call abslib,$(OUTDIR)/libxsmmgen.$(ILIBEXT)) \
+       $(LD) -o $@ $(OBJFILES_GEN_GEMM_BIN) -Wl,--rpath $(PREFIX)/lib lib/libxsmmgen.$(ILIBEXT) \
                $(call cleanld,$(NOBLAS_LDFLAGS) $(NOBLAS_CLDFLAGS))
 
 ifneq (,$(strip $(LIBJITPROFILING)))
diff --git a/Makefile.inc b/Makefile.inc
index 424d7b2e4..87934fee0 100644
--- a/Makefile.inc
+++ b/Makefile.inc
@@ -918,7 +918,7 @@ ifneq (Darwin,$(UNAME))
   XLNKVERBOSE := --verbose
   linkopt = $(if $1,$(XLNKOPT)$(if $2,$1=$(call quote,$2),$1))
   abslibrpath = $(strip $(if $(findstring .$(ILIBEXT),$1)$(wildcard $1/), \
-    $(call linkopt,--rpath,$(call qxdir,$(call qapath,$1)))))
+    $(call linkopt,--rpath,$(PREFIX)/lib)))
   XGROUP_BEGIN := $(call linkopt,--start-group)
   XGROUP_END := $(call linkopt,--end-group)
   ifneq (0,$(ASNEEDED))
```

File: pkgs/development/libraries/libxsmm/default.nix [hunk 2]
```diff
@@ -35,10 +35,23 @@ stdenv.mkDerivation rec {
...
+
     moveToOutput "share/libxsmm" "$doc"
```
Comment (markuskowa) [inline pkgs/development/libraries/libxsmm/default.nix:55]:
```suggestion
    moveToOutput "share/libxsmm" '${!outputDoc}"
```
(is the proper method, as I recently learned myself)

File: pkgs/development/libraries/libxsmm/default.nix (hunk 1 above)
Comment (sheepforce) [reply pkgs/development/libraries/libxsmm/default.nix]: Had to adapt a little bit for release 1.17 and not master, but yes, this is much better.

Comment (sheepforce) [reply pkgs/development/libraries/libxsmm/default.nix]: Yes that one works and finds all libraries. The patched line in `Makefile` does not exist on 1.17. But admittetdly, when playing around with that Git Repo, it behaved somewhat weird. Had problems checking out tags.

**Review by markuskowa — APPROVED** [review]

**Pull request merged**